app/src/main/java/com/kodeco/android/countryinfo/repository/CountryRepositoryImpl.kt

```diff
@@ -1,54 +1,60 @@
 package com.kodeco.android.countryinfo.repository
 
 import com.kodeco.android.countryinfo.model.Country
-import com.kodeco.android.countryinfo.networking.RemoteApiService
-import kotlinx.coroutines.flow.MutableStateFlow
-import kotlinx.coroutines.flow.asStateFlow
+import com.kodeco.android.countryinfo.repository.local.CountryLocalDataSource
+import com.kodeco.android.countryinfo.repository.remote.CountryRemoteDataSource
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
 
 class CountryRepositoryImpl(
-    private val apiService: RemoteApiService,
+    private val remoteDataSource: CountryRemoteDataSource,
+    private val localDataSource: CountryLocalDataSource,
 ) : CountryRepository {
 
-    private val _countries = MutableStateFlow(emptyList<Country>())
-    override val countries = _countries.asStateFlow()
+    private var _countries = mutableListOf<Country>()
 
-    private var favorites = setOf<String>()
+    override fun fetchCountries(): Flow<List<Country>> = flow {
```

Comment:
Continuing on here based on the changes in the CountryRepository interface you could do:

```kotlin
    override fun getCountries(): Flow<List<Country>> = flow {
        // Note: You'd need to pass in countryPrefs to the constructor of the CountryRepositoryImpl
        val localStorageEnabled = countryPrefs.getLocalStorageEnabled().first()
        val countries = if (localStorageEnabled) {
            getCountriesBasedOnSettings(localStorageEnabled)
        } else _countries

        // Store the favorited country names, so we can set the isFavorite value back after fetching from network
        val favoriteCountryNames = countries.filter { it.isFavorite }.map { it.commonName }

        val countriesResponse = remoteDataSource.fetchCountries()

        if (countriesResponse.isSuccessful) {
            countriesResponse.body()?.let {
                _countries.clear()
                _countries.addAll(it.map { country ->
                    if (favoriteCountryNames.contains(country.commonName)) {
                        country.copy(isFavorite = true)
                    } else country
                })

                if (localStorageEnabled) {
                    deleteAllCountries()
                    saveCountries(_countries)
                }
            }
        } else {
            // We don't have to throw here but we could use this else to log an error
//            throw Throwable("Request failed: ${countriesResponse.message()}")
        }

        emit(getCountriesBasedOnSettings(localStorageEnabled))
    }
    ...
    private suspend fun getCountriesBasedOnSettings(localStorageEnabled: Boolean): List<Country> {
        val favoritesFeatureEnabled = countryPrefs.getFavoritesFeatureEnabled().first()
        val favoriteCountriesEnabled = countryPrefs.getFavoriteCountriesFeatureEnabled().first()

        return if (localStorageEnabled) {
            if (favoritesFeatureEnabled && favoriteCountriesEnabled) {
                getLocalFavoriteCountries()
            } else {
                getLocalCountries()
            }
        } else _countries
    }

    private suspend fun getLocalCountries() = localDataSource.getAllCountries()

    private suspend fun getLocalFavoriteCountries() = localDataSource.getFavoriteCountries()
```